File: src/main/scala/net/machinemuse/numina/network/message/MusePacketModeChangeRequest.scala
package net.machinemuse.numina.network.message

import cpw.mods.fml.common.network.simpleimpl.{MessageContext, IMessageHandler, IMessage}
import io.netty.buffer.ByteBuf
import net.machinemuse.numina.item.ModeChangingItem
import net.machinemuse.numina.scala.OptionCast
import net.minecraft.entity.player.EntityPlayerMP

class MusePacketModeChangeRequest(slot: Int, mode: String) extends IMessage with IMessageHandler[MusePacketModeChangeRequest, IMessage] {
  var slotNew: Int = 0
  var modeNew: String = ""

  override def fromBytes(buf: ByteBuf) = {
    slotNew = buf.readInt()
    val length = buf.readInt()
    modeNew = new String(buf.readBytes(length).array())
  }

  override def toBytes(buf: ByteBuf) = {
    buf.writeInt(slot)
    buf.writeInt(mode.length)
    buf.writeBytes(mode.getBytes)
  }

  override def onMessage(message: MusePacketModeChangeRequest, ctx: MessageContext): IMessage = {
    val player: EntityPlayerMP = ctx.getServerHandler.playerEntity
    if (slotNew > -1 && slotNew < 9) {
      for {
        stack <- Option(player.inventory.mainInventory(slotNew))
        item <- OptionCast[ModeChangingItem](stack.getItem)
      } {
        val modes = item.getValidModes(stack, player)
        if (modes.contains(modeNew)) {
          OptionCast[ModeChangingItem](stack.getItem).map(i => i.setActiveMode(stack, modeNew))
        }
      }
    }
    null
  }
}

File: src/main/scala/net/machinemuse/numina/network/message/MusePacketModeChangeRequest.java
package net.machinemuse.numina.network.message;

import net.machinemuse.numina.item.ModeChangingItem;
import net.minecraft.item.ItemStack;

import cpw.mods.fml.common.network.ByteBufUtils;
import cpw.mods.fml.common.network.simpleimpl.IMessage;
import cpw.mods.fml.common.network.simpleimpl.IMessageHandler;
import cpw.mods.fml.common.network.simpleimpl.MessageContext;
import io.netty.buffer.ByteBuf;

public class MusePacketModeChangeRequest implements IMessage, IMessageHandler<MusePacketModeChangeRequest, IMessage> {
    private int slot;
    private String mode;

    public MusePacketModeChangeRequest() {}

    public MusePacketModeChangeRequest(int slot, String mode) {
        this.slot = slot;
        this.mode = mode;
    }

    @Override
    public void fromBytes(ByteBuf buf) {
        slot = buf.readInt();
        mode = ByteBufUtils.readUTF8String(buf);
    }

    @Override
    public void toBytes(ByteBuf buf) {
        buf.writeInt(slot);
        ByteBufUtils.writeUTF8String(buf, mode);
    }

    @Override
    public IMessage onMessage(MusePacketModeChangeRequest message, MessageContext ctx) {
        if (message.slot > -1 && message.slot < 9) {
            ItemStack stack = ctx.getServerHandler().playerEntity.inventory.mainInventory[message.slot];
            ModeChangingItem item = (ModeChangingItem)stack.getItem();
            if(item.getValidModes(stack, ctx.getServerHandler().playerEntity).contains(message.mode)) {
                item.setActiveMode(stack, message.mode);
            }
        }
        return null;
    }
}
